### PR TITLE
Add host_tx_ready enhancements

### PR DESCRIPTION
**Pull request by noaOrMlnx**
Depending on https://github.com/sonic-net/sonic-swss-common/pull/824
Depending on https://github.com/sonic-net/sonic-sairedis/pull/1307

HLD: https://github.com/sonic-net/SONiC/pull/1453

**What I did**
Add host_tx_ready enhancements to sonic-swss branch according to HLD mentioned above. 

**Why I did it**
Module and Asic configurations must be synchronized in order to not configure an non-existing module.

**How I verified it**
Manually & unit tests.

**Details if related**

### CODE REVIEW

**Review by Junchao-Mellanox — APPROVED** [review]

File: tests/test_warm_reboot.py [hunk 1]
```diff
@@ -60,6 +60,19 @@ def check_port_oper_status(appl_db, port_name, state):
             break
     assert oper_status == state
 
+def check_port_host_tx_ready_status(state_db, port_name, status):
```
Comment (prgeor) [inline tests/test_warm_reboot.py:63]: @noaOrMlnx have you tested in actual switch host_tx_ready is not deleted during warm-reboot? If its not delted, what is preventing it from deleted from STATE_DB's port table? Any code pointers?

File: orchagent/portsorch.cpp [hunk 2]
```diff
@@ -7345,6 +7521,18 @@ void PortsOrch::doTask(NotificationConsumer &consumer)
 
         sai_deserialize_free_port_oper_status_ntf(count, portoperstatus);
     }
+    else if (&consumer == m_portHostTxReadyNotificationConsumer && op == "port_host_tx_ready")
```
Comment (prgeor) [inline orchagent/portsorch.cpp:7524]: @noaOrMlnx Have you tested, when user does admin down, the host_tx_ready is false?

File: orchagent/portsorch.cpp [hunk 3]
```diff
@@ -4469,6 +4561,69 @@ void PortsOrch::doVlanMemberTask(Consumer &consumer)
     }
 }
 
+void PortsOrch::doTransceiverInfoTableTask(Consumer &consumer)
```
Comment (prgeor) [inline orchagent/portsorch.cpp:4564]: @noaOrMlnx I would suggest to rename "doTransceiverPresenceCheck" since this routine is checking the presence of the module as notified in TRANSCEIVER_INFO SET/DEL

File: orchagent/portsorch.cpp [hunk 4]
```diff
@@ -4469,6 +4561,69 @@ void PortsOrch::doVlanMemberTask(Consumer &consumer)
...
+        }
+
+        it = consumer.m_toSync.erase(it);
+    }
+}
+
+bool PortsOrch::setSaiHostTxSignal(const Port &port, bool enable)
+{
+    sai_attribute_t attr;
+    attr.id = SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE;
```
Comment (prgeor) [inline orchagent/portsorch.cpp:4614]: @noaOrMlnx Will this crash for platform that doesn't support this new SAI attribute? See failure message?

File: orchagent/portsorch.cpp (hunk 2 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp:7524]: Yes

File: tests/test_warm_reboot.py (hunk 1 above)
Comment (noaOrMlnx) [reply tests/test_warm_reboot.py:63]: The table is deleted. 
The thing is, that after warm reboot, host_tx_ready value will be false, and we won't get any new notification from lower layers saying that host_tx_ready=true. 
That's the reason we need to take the current value of host_tx_ready using the new API

**Review by prgeor — APPROVED** [review]

**Comment by prgeor** [comment]
@noaOrMlnx please check if failure is due to some other PR dependency

**Review by prgeor — DISMISSED** [review]

**Review by prgeor — APPROVED** [review]

File: orchagent/main.cpp [hunk 5]
```diff
@@ -505,6 +505,10 @@ int main(int argc, char **argv)
     attr.value.ptr = (void *)on_switch_shutdown_request;
     attrs.push_back(attr);
 
+    attr.id = SAI_SWITCH_ATTR_PORT_HOST_TX_READY_NOTIFY;
```
Comment (prsunny) [inline orchagent/main.cpp:508]: Does it handle if a platform does not support this notification?

File: orchagent/orchdaemon.cpp [hunk 6]
```diff
@@ -174,7 +174,7 @@ bool OrchDaemon::init()
     gDirectory.set(vnet_rt_orch);
     VRFOrch *vrf_orch = new VRFOrch(m_applDb, APP_VRF_TABLE_NAME, m_stateDb, STATE_VRF_OBJECT_TABLE_NAME);
     gDirectory.set(vrf_orch);
-    gMonitorOrch = new MonitorOrch(m_stateDb, STATE_VNET_MONITOR_TABLE_NAME);
+    gMonitorOrch = new MonitorOrch(m_stateDb, STATE_VNET_MONITOR_TABLE_NAME);
```
Comment (prsunny) [inline orchagent/orchdaemon.cpp:177]: What is this change? Could you revert?

File: orchagent/portsorch.cpp [hunk 7]
```diff
@@ -391,6 +391,8 @@ static void getPortSerdesAttr(PortSerdesAttrMap_t &map, const PortConfig &port)
  *    bridge. By design, SONiC switch starts with all bridge ports removed from
  *    default VLAN and all ports removed from .1Q bridge.
  */
+
```
Comment (prsunny) [inline orchagent/portsorch.cpp:394]: I suggest, please revert changes that are not related to this PR.

File: orchagent/portsorch.cpp [hunk 8]
```diff
@@ -5157,6 +5316,7 @@ bool PortsOrch::initializePort(Port &port)
         port.m_oper_status = SAI_PORT_OPER_STATUS_DOWN;
     }
 
+
```
Comment (prsunny) [inline orchagent/portsorch.cpp:5319]: Extra line

File: orchagent/main.cpp (hunk 5 above)
Comment (noaOrMlnx) [reply orchagent/main.cpp:508]: Sure, this line purpose is only to have the callback function for the notification. 
In PortsOrch we have 2 queries to check whether this functionality is supported, and if not, the notification will not be registered and this line will not be relevant.

File: orchagent/orchdaemon.cpp (hunk 6 above)
Comment (noaOrMlnx) [reply orchagent/orchdaemon.cpp:177]: Done

File: orchagent/portsorch.cpp (hunk 7 above)
Comment (prsunny) [reply orchagent/portsorch.cpp:394]: I still see extra lines added. Could you please address the comment?

**Review by prsunny — DISMISSED** [review]

File: orchagent/portsorch.cpp [hunk 9]
```diff
@@ -1441,6 +1499,20 @@ bool PortsOrch::setPortAdminStatus(Port &port, bool state)
...
+{
+    Port p;
+
+    if (!getPort(portId, p))
+    {
+        SWSS_LOG_ERROR("Failed to get port object for port id 0x%" PRIx64, portId);
+        return;
+    }
+
+    SWSS_LOG_NOTICE("Setting host_tx_ready status = %s, port_id = 0x%" PRIx64, status.c_str(), portId);
```
Comment (prgeor) [inline orchagent/portsorch.cpp:1512]: @noaOrMlnx please log the logical port name (port.m_alias) as well for easy identification

File: orchagent/portsorch.cpp [hunk 10]
```diff
@@ -1423,15 +1481,15 @@ bool PortsOrch::setPortAdminStatus(Port &port, bool state)
...
+    if (gbstatus != true && !m_cmisModuleAsicSyncSupported)
     {
         m_portStateTable.hset(port.m_alias, "host_tx_ready", "false");
         SWSS_LOG_NOTICE("Set host_tx_ready to false as gbstatus is false "
                         "for port %s", port.m_alias.c_str());
     }
 
     /* Update the state table for host_tx_ready*/
-    if (state && (gbstatus == true) && (status == SAI_STATUS_SUCCESS) )
+    if (state && (gbstatus == true) && (status == SAI_STATUS_SUCCESS) && !m_cmisModuleAsicSyncSupported)
```
Comment (prgeor) [inline orchagent/portsorch.cpp:1492]: @noaOrMlnx can we use the newly introduced API setHostTxReady() here as well instead of directly calling hset?

File: orchagent/portsorch.cpp [hunk 11]
```diff
@@ -1414,7 +1468,11 @@ bool PortsOrch::setPortAdminStatus(Port &port, bool state)
         SWSS_LOG_ERROR("Failed to set admin status %s for port %s."
                        " Setting host_tx_ready as false",
                        state ? "UP" : "DOWN", port.m_alias.c_str());
-        m_portStateTable.hset(port.m_alias, "host_tx_ready", "false");
+
+        if (!m_cmisModuleAsicSyncSupported)
+        {
+            m_portStateTable.hset(port.m_alias, "host_tx_ready", "false");
```
Comment (prgeor) [inline orchagent/portsorch.cpp:1474]: @noaOrMlnx can we use the newly introduced API setHostTxReady() here as well instead of directly calling hset?

File: orchagent/portsorch.cpp [hunk 12]
```diff
@@ -1400,8 +1451,11 @@ bool PortsOrch::setPortAdminStatus(Port &port, bool state)
     attr.id = SAI_PORT_ATTR_ADMIN_STATE;
     attr.value.booldata = state;
 
+    // if sync between cmis module configuration and asic is supported,
+    // do not change host_tx_ready value in STATE DB when admin status is changed.
+
     /* Update the host_tx_ready to false before setting admin_state, when admin state is false */
-    if (!state)
+    if (!state && !m_cmisModuleAsicSyncSupported)
```
Comment (prgeor) [inline orchagent/portsorch.cpp:1458]: @noaOrMlnx can we use the newly introduced API setHostTxReady() here as well instead of directly calling hset?

**Comment by prsunny** [comment]
@noaOrMlnx , this is ready for merge.  Waiting for @prgeor to signoff.

File: orchagent/portsorch.cpp (hunk 9 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp:1512]: Done

File: orchagent/portsorch.cpp (hunk 10 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp:1492]: Done

File: orchagent/portsorch.cpp (hunk 11 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp:1474]: Done

File: orchagent/portsorch.cpp (hunk 12 above)
Comment (noaOrMlnx) [reply orchagent/portsorch.cpp:1458]: Done

**Review by prsunny — APPROVED** [review]

File: orchagent/portsorch.cpp (hunk 9 above)
Comment (prgeor) [reply orchagent/portsorch.cpp:1511]: @noaOrMlnx please add the logical port name as done previously `port.m_alias.c_Str()`

Comment (noaOrMlnx) [reply orchagent/portsorch.cpp]: @prgeor Done

**Review by prgeor — APPROVED** [review]

**Review by prsunny — APPROVED** [review]
